File: pkg/node/js/zlib.js
// zlib implements node:zlib on top of one Go host call that runs a codec end to
// end and carries both sides as base64. It covers the three framings the Go
// standard library implements, zlib (deflate), raw deflate and gzip, in all
// three shapes Node offers them: the *Sync calls, the callback forms, and the
// Transform streams.
//
// Two things Node has are not here. Brotli would need a compressor outside the
// standard library, and bento links no C, so brotliCompress and its siblings are
// absent rather than present and throwing: code that feature-detects them then
// takes its non-brotli path. And the streams buffer their whole input and run
// the codec once at flush, rather than compressing incrementally, so a stream
// over a large file holds it in memory and emits one chunk. What comes out is
// the same bytes either way; what differs is when, and how much is resident.

__bento_defineModule("zlib", function (module, exports, require) {
  "use strict";

  const { Buffer } = require("buffer");
  const { Transform } = require("stream");

  // The subset of zlib's constants a caller actually passes or compares
  // against. Node exposes these both on the module and under .constants, and
  // programs read them from either place.
  const constants = {
    Z_NO_FLUSH: 0,
    Z_PARTIAL_FLUSH: 1,
    Z_SYNC_FLUSH: 2,
    Z_FULL_FLUSH: 3,
    Z_FINISH: 4,
    Z_BLOCK: 5,
    Z_TREES: 6,

    Z_OK: 0,
    Z_STREAM_END: 1,
    Z_NEED_DICT: 2,
    Z_ERRNO: -1,
    Z_STREAM_ERROR: -2,
    Z_DATA_ERROR: -3,
    Z_MEM_ERROR: -4,
    Z_BUF_ERROR: -5,
    Z_VERSION_ERROR: -6,

    Z_NO_COMPRESSION: 0,
    Z_BEST_SPEED: 1,
    Z_BEST_COMPRESSION: 9,
    Z_DEFAULT_COMPRESSION: -1,

    Z_FILTERED: 1,
    Z_HUFFMAN_ONLY: 2,
    Z_RLE: 3,
    Z_FIXED: 4,
    Z_DEFAULT_STRATEGY: 0,

    DEFLATE: 1,
    INFLATE: 2,
    GZIP: 3,
    GUNZIP: 4,
    DEFLATERAW: 5,
    INFLATERAW: 6,
    UNZIP: 7,
  };

  // The numbers zlib returns for the codes the host can report, so err.errno
  // matches err.code the way it does in Node.
  const ERRNO = {
    Z_DATA_ERROR: constants.Z_DATA_ERROR,
    Z_BUF_ERROR: constants.Z_BUF_ERROR,
    Z_STREAM_ERROR: constants.Z_STREAM_ERROR,
  };

  // Node accepts a string, a Buffer, an ArrayBuffer or any view over one. A
  // string is utf8, which is what Buffer.from defaults to.
  function toBuffer(data) {
    if (Buffer.isBuffer(data)) return data;
    if (typeof data === "string") return Buffer.from(data, "utf8");
    if (data instanceof ArrayBuffer) return Buffer.from(new Uint8Array(data));
    if (ArrayBuffer.isView(data)) return Buffer.from(new Uint8Array(data.buffer, data.byteOffset, data.byteLength));
    throw new TypeError("The \"buffer\" argument must be of type string or an instance of Buffer, TypedArray, DataView, or ArrayBuffer");
  }

  function levelOf(opts) {
    if (opts && typeof opts.level === "number") return opts.level;
    return constants.Z_DEFAULT_COMPRESSION;
  }

  function run(format, direction, data, opts) {
    const input = toBuffer(data);
    const env = JSON.parse(__bento_zlib_run(format, direction, input.toString("base64"), levelOf(opts)));
    if (!env.ok) {
      const err = new Error(env.msg || "zlib error");
      err.code = env.code || "Z_DATA_ERROR";
      err.errno = ERRNO[err.code] !== undefined ? ERRNO[err.code] : constants.Z_ERRNO;
      throw err;
    }
    return Buffer.from(env.b64 || "", "base64");
  }

  // The callback forms take (buffer, [options], callback) and, like Node's, are
  // asynchronous even though the work underneath is not: a caller that passes a
  // callback expects it after the current turn, not during the call.
  function async(format, direction) {
    return function (data, opts, cb) {
      if (typeof opts === "function") {
        cb = opts;
        opts = undefined;
      }
      if (typeof cb !== "function") throw new TypeError("Callback must be a function");
      let result, error;
      try {
        result = run(format, direction, data, opts);
      } catch (e) {
        error = e;
      }
      Promise.resolve().then(function () {
        if (error) cb(error);
        else cb(null, result);
      });
    };
  }

  // Zlib is the Transform every codec stream is. It collects the chunks written
  // to it and runs the codec once, on flush: see the note at the top of the file
  // for what that costs and what it does not change.
  class Zlib extends Transform {
    constructor(format, direction, opts) {
      super(opts);
      this._format = format;
      this._direction = direction;
      this._opts = opts || {};
      this._chunks = [];
      this.bytesWritten = 0;
    }
    _transform(chunk, encoding, callback) {
      const buf = toBuffer(chunk);
      this.bytesWritten += buf.length;
      this._chunks.push(buf);
      callback();
    }
    _flush(callback) {
      let out;
      try {
        out = run(this._format, this._direction, Buffer.concat(this._chunks), this._opts);
      } catch (e) {
        callback(e);
        return;
      }
      this.push(out);
      callback();
    }
  }

  // Each codec gets the four names Node gives it: the class, its factory, the
  // synchronous call and the callback form.
  const CODECS = [
    { name: "Deflate", fn: "deflate", format: "deflate", direction: "compress" },
    { name: "Inflate", fn: "inflate", format: "deflate", direction: "decompress" },
    { name: "DeflateRaw", fn: "deflateRaw", format: "deflateRaw", direction: "compress" },
    { name: "InflateRaw", fn: "inflateRaw", format: "deflateRaw", direction: "decompress" },
    { name: "Gzip", fn: "gzip", format: "gzip", direction: "compress" },
    { name: "Gunzip", fn: "gunzip", format: "gzip", direction: "decompress" },
    // unzip is not a fourth framing but inflate's sniffing mode: it takes either
    // a gzip member or a zlib stream and decides by the header.
    { name: "Unzip", fn: "unzip", format: "unzip", direction: "decompress" },
  ];

  const zlib = { constants: constants };
  for (const key of Object.keys(constants)) zlib[key] = constants[key];

  for (const codec of CODECS) {
    const cls = class extends Zlib {
      constructor(opts) {
        super(codec.format, codec.direction, opts);
      }
    };
    Object.defineProperty(cls, "name", { value: codec.name });
    zlib[codec.name] = cls;
    zlib["create" + codec.name] = (opts) => new cls(opts);
    zlib[codec.fn + "Sync"] = (data, opts) => run(codec.format, codec.direction, data, opts);
    zlib[codec.fn] = async(codec.format, codec.direction);
  }

  module.exports = zlib;
});

File: pkg/node/zlib.go
package node

import (
	"bytes"
	"compress/flate"
	"compress/gzip"
	"compress/zlib"
	"encoding/base64"
	"errors"
	"io"
)

// zlibResult is the JSON envelope the compression host function returns. It
// mirrors fsResult's shape: a success flag, the payload as base64, and on
// failure the code a program branches on plus the text zlib would have put in
// the message.
type zlibResult struct {
	OK   bool   `json:"ok"`
	B64  string `json:"b64,omitempty"`
	Code string `json:"code,omitempty"`
	Msg  string `json:"msg,omitempty"`
}

// The three wire formats Go's standard library covers, named as node:zlib names
// them. deflate is a zlib stream (RFC 1950), deflateRaw is a bare deflate
// stream (RFC 1951), and gzip is a gzip member (RFC 1952). unzip is not a
// format but inflate's sniffing mode: it accepts either of the first two
// framings and picks by the header.
const (
	formatDeflate    = "deflate"
	formatDeflateRaw = "deflateRaw"
	formatGzip       = "gzip"
	formatUnzip      = "unzip"
)

// zlibHostFuncs exposes one call for every codec. The direction, the framing and
// the level all ride as arguments rather than as separate functions, because the
// work either side of the compress/decompress split is the same marshalling.
func zlibHostFuncs() map[string]HostFunc {
	return map[string]HostFunc{
		"__bento_zlib_run": hostZlibRun,
	}
}

// hostZlibRun takes (format, direction, base64 input, level) and returns the
// envelope. Level is only read when compressing, and -1 asks for the library
// default.
func hostZlibRun(a []any) (any, error) {
	format := str(a, 0)
	direction := str(a, 1)
	data, err := base64.StdEncoding.DecodeString(str(a, 2))
	if err != nil {
		return zlibFail("Z_DATA_ERROR", "invalid input encoding"), nil
	}
	level := flate.DefaultCompression
	if len(a) > 3 && a[3] != nil {
		level = intArg(a, 3)
	}

	var out []byte
	if direction == "compress" {
		out, err = zlibCompress(format, data, level)
	} else {
		out, err = zlibDecompress(format, data)
	}
	if err != nil {
		return zlibFail(zlibCode(err), err.Error()), nil
	}
	return jsonString(zlibResult{OK: true, B64: base64.StdEncoding.EncodeToString(out)}), nil
}

func zlibCompress(format string, data []byte, level int) ([]byte, error) {
	// Go rejects any level outside [-2, 9]; zlib clamps instead, and Node passes
	// whatever the caller wrote straight through, so clamp here to keep a level of
	// 42 from turning into an error the C library would never raise.
	if level < flate.HuffmanOnly {
		level = flate.HuffmanOnly
	}
	if level > flate.BestCompression {
		level = flate.BestCompression
	}

	var buf bytes.Buffer
	var w io.WriteCloser
	var err error
	switch format {
	case formatDeflate:
		w, err = zlib.NewWriterLevel(&buf, level)
	case formatDeflateRaw:
		w, err = flate.NewWriter(&buf, level)
	case formatGzip:
		w, err = gzip.NewWriterLevel(&buf, level)
	default:
		return nil, errors.New("unknown compression format: " + format)
	}
	if err != nil {
		return nil, err
	}
	if _, err := w.Write(data); err != nil {
		w.Close()
		return nil, err
	}
	// The trailer only lands on Close, so the buffer is not the stream until then.
	if err := w.Close(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

func zlibDecompress(format string, data []byte) ([]byte, error) {
	if format == formatUnzip {
		format = sniff(data)
	}
	// The header is checked before the stream is read so that input which is not
	// the framing at all is reported as a data error rather than as a truncated
	// one. Go's readers do not distinguish: a two-byte scrap and a genuine gzip
	// member cut short both come back as an unexpected EOF, and zlib calls only
	// the second of those Z_BUF_ERROR.
	if err := checkHeader(format, data); err != nil {
		return nil, err
	}
	switch format {
	case formatDeflate:
		return readAllClosing(func() (io.ReadCloser, error) { return zlib.NewReader(bytes.NewReader(data)) })
	case formatDeflateRaw:
		return readAllClosing(func() (io.ReadCloser, error) { return flate.NewReader(bytes.NewReader(data)), nil })
	case formatGzip:
		// Node's gunzip reads every member of a multi-member file, which is what
		// concatenated gzip output is; Go does the same once multistream is left on.
		return readAllClosing(func() (io.ReadCloser, error) { return gzip.NewReader(bytes.NewReader(data)) })
	default:
		return nil, errors.New("unknown compression format: " + format)
	}
}

// errBadHeader is the failure for input whose framing is wrong from the first
// bytes. zlib reports it as "incorrect header check", and reports it even for
// input too short to hold a whole header, which is why this is decided here
// rather than left to the reader.
var errBadHeader = errors.New("incorrect header check")

// checkHeader validates the fixed prefix of a framing. Raw deflate has no
// header to check, so nothing is rejected there.
func checkHeader(format string, data []byte) error {
	switch format {
	case formatGzip:
		if len(data) < 2 || data[0] != 0x1f || data[1] != 0x8b {
			return errBadHeader
		}
	case formatDeflate:
		// RFC 1950: the low nibble of the first byte is the compression method,
		// which is 8 for deflate, and the two header bytes read as a big-endian
		// number are a multiple of 31.
		if len(data) < 2 || data[0]&0x0f != 8 || (int(data[0])<<8|int(data[1]))%31 != 0 {
			return errBadHeader
		}
	}
	return nil
}

// sniff picks the framing for unzip. A gzip member starts with the fixed magic
// 1f 8b; a zlib stream starts with a compression-method-and-flags byte whose low
// nibble is 8 for deflate, and whose first two bytes are a multiple of 31. What
// matches neither is treated as a raw deflate stream, which is the only framing
// left.
func sniff(data []byte) string {
	if len(data) >= 2 && data[0] == 0x1f && data[1] == 0x8b {
		return formatGzip
	}
	if len(data) >= 2 && data[0]&0x0f == 8 && (int(data[0])<<8|int(data[1]))%31 == 0 {
		return formatDeflate
	}
	return formatDeflateRaw
}

// readAllClosing runs a reader to completion and closes it. The constructor is a
// callback rather than a reader because zlib and gzip validate their header when
// they are built, and that failure has to be reported like any other.
func readAllClosing(open func() (io.ReadCloser, error)) ([]byte, error) {
	r, err := open()
	if err != nil {
		return nil, err
	}
	out, err := io.ReadAll(r)
	closeErr := r.Close()
	if err != nil {
		return nil, err
	}
	if closeErr != nil {
		return nil, closeErr
	}
	return out, nil
}

// zlibCode maps a Go decompression failure onto the code node:zlib puts on the
// error. A stream that stops early is Z_BUF_ERROR, the way zlib reports running
// out of input; anything else malformed is Z_DATA_ERROR.
func zlibCode(err error) string {
	if errors.Is(err, errBadHeader) {
		return "Z_DATA_ERROR"
	}
	if errors.Is(err, io.ErrUnexpectedEOF) || errors.Is(err, io.EOF) {
		return "Z_BUF_ERROR"
	}
	return "Z_DATA_ERROR"
}

func zlibFail(code, msg string) string {
	return jsonString(zlibResult{OK: false, Code: code, Msg: msg})
}
